src/Animate/Animate.man
.\" t
.\" @(#)Animate.1	02/07/00
.TH Animate 1.8 "Feb 7 2000" Animate
.UC
.SH NAME
\fBAnimate\fP \- the AfterStep iconify animation module

.SH SYNOPSIS
\fBAnimate\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
.sp
Although \fBAnimate\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
	Module	"I"	Animate
.sp
.fi
or from within an afterstep menu file by 'Animate':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Animate"
	Module "Animate" Animate
	MiniPixmap "mini-as.xpm"
.fi

To stop the \fBAnimate\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Animate_Stop"
	KillModuleByName "Stop Animate" Animate
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBAnimate\fP module performs animation on window iconification and
deiconification in several styles and colors.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fBAnimate\fP will search for a configuration file
which describes the colors, style and actions to take.  This file will be 
the one specified by the \fI\-f\fP option (if given).  Else, the 
~/GNUstep/Library/AfterStep/animate file will be used if it exists.  Else, 
/usr/local/share/afterstep/animate will be used.

.SH INVOCATION
\fBAnimate\fP can be invoked from the \fIautoexec\fP file or from a menu
item as shown in the Synopsis section above.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate Animate.

.SH CONFIGURATION FILE
\fBAnimate\fP reads the \fIanimate\fP file when it starts up and looks
for it's configuration options:

.IP "*AnimateColor \fBcolor\fP"
Tells Animate what color (see rgb.txt) to draw the window outline with as
it animates.

.IP "*AnimateDelay \fBmsecs\fP"
Tells Animate how many milliseconds to sleep between frames of animation.

.IP "*AnimateIterations \fBiterations\fP"
Tells Animate how many steps to break the animation into.

.IP "*AnimateTwist \fBtwist\fP"
Tells Animate how many revolutions to twist the iconification frame.

.IP "*AnimateWidth \fBwidth\fP"
Tells Animate how wide a line to draw with.  Note that with 0 produces a
fast line at Width 1.

.IP "*AnimateResize \fBmode\fP"
Tells Animate what method to use to draw resizes.  \fIzoom\fP,  \fIzoom3D\fP,
\fItwist\fP, \fIflip\fP, \fIturn\fP & \fIrandom\fP are currently supported.


.SH FILES

.IP \fI/usr/local/share/afterstep/animate\fP
The global configuration file for system wide defaults of all users.

.IP \fI$HOME/GNUstep/Library/AfterStep/animate\fP
This file should be copied from the \fI/usr/local/share/afterstep/animate\fP and edited to suit a user's specific requirements.  This file, if it exists,
will override the system wide default file.

.SH EXAMPLES
The example is the \fIanimate\fP file shipped, refer to the \fBFILES\fP
section above for its configuration options.

.SH AUTHORS
Alfredo Kengi Kojima <kojima@inf.ufrgs.br>
.nf
Kaj Groner <kajg@mindspring.com>
.nf
Frank Scheelen <scheelen@worldonline.nl>
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1),
.BR Wharf (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha


src/Animate/Animate.c
/*
 * Copyright (c) 1997 Frank Scheelen <scheelen@worldonline.nl>
 * Copyright (c) 1996 Alfredo Kengi Kojima (kojima@inf.ufrgs.br)
 * Copyright (c) 1996 Kaj Groner <kajg@mindspring.com>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <ctype.h>
#include <errno.h>
#include <math.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

#include <limits.h>
#include <unistd.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#define MODULE_X_INTERFACE

#include "../../configure.h"
#include "../../include/aftersteplib.h"
#include "../../include/afterstep.h"
#include "../../include/style.h"
#include "../../include/screen.h"
#include "../../include/module.h"
#include "Animate.h"

#define AS_PI 3.14159265358979323846

Display *dpy;
int screen;
char *MyName = NULL;

ScreenInfo Scr;			/* root window */
GC DrawGC;
char *MyName;

/* File type information */
static int fd_width;
static int fd[2];
static int x_fd;

/* masks for AS pipe */
#define mask_reg (	M_CONFIGURE_WINDOW |	\
			M_ICONIFY |		\
			M_DEICONIFY |		\
			M_LOCKONSEND	)
#define mask_off 0

void Loop ();
void ParseOptions (const char *);

struct ASAnimate Animate =
{NULL, ANIM_ITERATIONS, ANIM_DELAYMS,
 ANIM_TWIST, ANIM_WIDTH, AnimateResizeTwist};

/*
 * This makes a twisty iconify/deiconify animation for a window, similar to
 * MacOS.  Parameters specify the position and the size of the initial
 * window and the final window
 */
void
AnimateResizeTwist (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  float cx, cy, cw, ch;
  float xstep, ystep, wstep, hstep;
  XPoint points[5];
  float angle, angle_finite, a, d;

  x += w / 2;
  y += h / 2;
  fx += fw / 2;
  fy += fh / 2;

  xstep = (float) (fx - x) / Animate.iterations;
  ystep = (float) (fy - y) / Animate.iterations;
  wstep = (float) (fw - w) / Animate.iterations;
  hstep = (float) (fh - h) / Animate.iterations;

  cx = (float) x;
  cy = (float) y;
  cw = (float) w;
  ch = (float) h;
  a = atan (ch / cw);
  d = sqrt ((cw / 2) * (cw / 2) + (ch / 2) * (ch / 2));

  angle_finite = 2 * AS_PI * Animate.twist;
  for (angle = 0;; angle += (float) (2 * AS_PI * Animate.twist / Animate.iterations))
    {
      if (angle > angle_finite)
	angle = angle_finite;
      points[0].x = cx + cos (angle - a) * d;
      points[0].y = cy + sin (angle - a) * d;
      points[1].x = cx + cos (angle + a) * d;
      points[1].y = cy + sin (angle + a) * d;
      points[2].x = cx + cos (angle - a + AS_PI) * d;
      points[2].y = cy + sin (angle - a + AS_PI) * d;
      points[3].x = cx + cos (angle + a + AS_PI) * d;
      points[3].y = cy + sin (angle + a + AS_PI) * d;
      points[4].x = cx + cos (angle - a) * d;
      points[4].y = cy + sin (angle - a) * d;
      XGrabServer (dpy);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XFlush (dpy);
      sleep_a_little (Animate.delay * 1000);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XUngrabServer (dpy);
      cx += xstep;
      cy += ystep;
      cw += wstep;
      ch += hstep;
      a = atan (ch / cw);
      d = sqrt ((cw / 2) * (cw / 2) + (ch / 2) * (ch / 2));
      if (angle >= angle_finite)
	break;
    }
  XFlush (dpy);
}

 /*
  * Add even more 3D feel to AfterStep by doing a flipping iconify.
  * Parameters specify the position and the size of the initial and the
  * final window.
  *
  * Idea: how about texture mapped, user definable free 3D movement
  * during a resize? That should get X on its knees all right! :)
  */
void
AnimateResizeFlip (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  float cx, cy, cw, ch;
  float xstep, ystep, wstep, hstep;
  XPoint points[5];

  float distortx;
  float distortch;
  float midy;

  float angle, angle_finite;

  xstep = (float) (fx - x) / Animate.iterations;
  ystep = (float) (fy - y) / Animate.iterations;
  wstep = (float) (fw - w) / Animate.iterations;
  hstep = (float) (fh - h) / Animate.iterations;

  cx = (float) x;
  cy = (float) y;
  cw = (float) w;
  ch = (float) h;

  angle_finite = 2 * AS_PI * Animate.twist;
  for (angle = 0;; angle += (float) (2 * AS_PI * Animate.twist / Animate.iterations))
    {
      if (angle > angle_finite)
	angle = angle_finite;

      distortx = (cw / 10) - ((cw / 5) * sin (angle));
      distortch = (ch / 2) * cos (angle);
      midy = cy + (ch / 2);

      points[0].x = cx + distortx;
      points[0].y = midy - distortch;
      points[1].x = cx + cw - distortx;
      points[1].y = points[0].y;
      points[2].x = cx + cw + distortx;
      points[2].y = midy + distortch;
      points[3].x = cx - distortx;
      points[3].y = points[2].y;
      points[4].x = points[0].x;
      points[4].y = points[0].y;

      XGrabServer (dpy);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XFlush (dpy);
      sleep_a_little (Animate.delay * 1000);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XUngrabServer (dpy);
      cx += xstep;
      cy += ystep;
      cw += wstep;
      ch += hstep;
      if (angle >= angle_finite)
	break;
    }
  XFlush (dpy);
}


 /*
  * And another one, this time around the Y-axis.
  */
void
AnimateResizeTurn (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  float cx, cy, cw, ch;
  float xstep, ystep, wstep, hstep;
  XPoint points[5];

  float distorty;
  float distortcw;
  float midx;

  float angle, angle_finite;

  xstep = (float) (fx - x) / Animate.iterations;
  ystep = (float) (fy - y) / Animate.iterations;
  wstep = (float) (fw - w) / Animate.iterations;
  hstep = (float) (fh - h) / Animate.iterations;

  cx = (float) x;
  cy = (float) y;
  cw = (float) w;
  ch = (float) h;

  angle_finite = 2 * AS_PI * Animate.twist;
  for (angle = 0;; angle += (float) (2 * AS_PI * Animate.twist / Animate.iterations))
    {
      if (angle > angle_finite)
	angle = angle_finite;

      distorty = (ch / 10) - ((ch / 5) * sin (angle));
      distortcw = (cw / 2) * cos (angle);
      midx = cx + (cw / 2);

      points[0].x = midx - distortcw;
      points[0].y = cy + distorty;
      points[1].x = midx + distortcw;
      points[1].y = cy - distorty;
      points[2].x = points[1].x;
      points[2].y = cy + ch + distorty;
      points[3].x = points[0].x;
      points[3].y = cy + ch - distorty;
      points[4].x = points[0].x;
      points[4].y = points[0].y;

      XGrabServer (dpy);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XFlush (dpy);
      sleep_a_little (Animate.delay * 1000);
      XDrawLines (dpy, Scr.Root, DrawGC, points, 5, CoordModeOrigin);
      XUngrabServer (dpy);
      cx += xstep;
      cy += ystep;
      cw += wstep;
      ch += hstep;
      if (angle >= angle_finite)
	break;
    }
  XFlush (dpy);
}


/*
 * This makes a zooming iconify/deiconify animation for a window, like most
 * any other icon animation out there.  Parameters specify the position and
 * the size of the initial window and the final window
 */
void
AnimateResizeZoom (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  float cx, cy, cw, ch;
  float xstep, ystep, wstep, hstep;
  int i;

  xstep = (float) (fx - x) / Animate.iterations;
  ystep = (float) (fy - y) / Animate.iterations;
  wstep = (float) (fw - w) / Animate.iterations;
  hstep = (float) (fh - h) / Animate.iterations;

  cx = (float) x;
  cy = (float) y;
  cw = (float) w;
  ch = (float) h;
  for (i = 0; i < Animate.iterations; i++)
    {
      XGrabServer (dpy);
      XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw, (int) ch);
      XFlush (dpy);
      sleep_a_little (Animate.delay);
      XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw, (int) ch);
      XUngrabServer (dpy);
      cx += xstep;
      cy += ystep;
      cw += wstep;
      ch += hstep;
    }
  XFlush (dpy);
}

/*
 * The effect of this is similar to AnimateResizeZoom but this time we
 * add lines to create a 3D effect.  The gotcha is that we have to do
 * something different depending on the direction we are zooming in.
 *
 * Andy Parker <parker_andy@hotmail.com>
 */
void
AnimateResizeZoom3D (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  float cx, cy, cw, ch;
  float xstep, ystep, wstep, hstep, srca, dsta;
  int i;

  xstep = (float) (fx - x) / Animate.iterations;
  ystep = (float) (fy - y) / Animate.iterations;
  wstep = (float) (fw - w) / Animate.iterations;
  hstep = (float) (fh - h) / Animate.iterations;
  dsta = (float) (fw + fh);
  srca = (float) (w + h);

  cx = (float) x;

  cy = (float) y;
  cw = (float) w;
  ch = (float) h;

  if (dsta <= srca)
    /* We are going from a Window to an Icon */
    {
      for (i = 0; i < Animate.iterations; i++)
	{
	  XGrabServer (dpy);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw,
			  (int) ch);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) fx, (int) fy, (int) fw,
			  (int) fh);
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, fx, fy);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw), (int) cy,
		     (fx + fw), fy);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw),
		     ((int) cy + (int) ch), (fx + fw), (fy + fh));
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, ((int) cy + (int) ch), fx,
		     (fy + fh));
	  XFlush (dpy);
	  sleep_a_little (Animate.delay);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw,
			  (int) ch);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) fx, (int) fy, (int) fw,
			  (int) fh);
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, fx, fy);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw), (int) cy,
		     (fx + fw), fy);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw),
		     ((int) cy + (int) ch), (fx + fw), (fy + fh));
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, ((int) cy + (int) ch), fx,
		     (fy + fh));
	  XUngrabServer (dpy);
	  cx += xstep;
	  cy += ystep;
	  cw += wstep;
	  ch += hstep;
	}
    }
  if (dsta > srca)
    {
/* We are going from an Icon to a Window */
      for (i = 0; i < Animate.iterations; i++)
	{
	  XGrabServer (dpy);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw,
			  (int) ch);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, w, h);
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, x, y);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw), (int) cy,
		     (x + w), y);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw), ((int) cy +
					       (int) ch), (x + w), (y + h));
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, ((int) cy + (int) ch), x,
		     (y + h));
	  XFlush (dpy);
	  sleep_a_little (Animate.delay);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, (int) cw,
			  (int) ch);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, w, h);
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, (int) cy, x, y);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw), (int) cy,
		     (x + w), y);
	  XDrawLine (dpy, Scr.Root, DrawGC, ((int) cx + (int) cw),
		     ((int) cy + (int) ch), (x + w), (y + h));
	  XDrawLine (dpy, Scr.Root, DrawGC, (int) cx, ((int) cy + (int) ch), x,
		     (y + h));
	  XUngrabServer (dpy);
	  cx += xstep;
	  cy += ystep;
	  cw += wstep;
	  ch += hstep;
	}
    }
  XFlush (dpy);
}

 /*
  * This picks one of the above animations.
  */
void
AnimateResizeRandom (int x, int y, int w, int h, int fx, int fy, int fw, int fh)
{
  switch ((rand () + (x * y + w * h + fx)) % 4)
    {
    case 0:
      AnimateResizeTwist (x, y, w, h, fx, fy, fw, fh);
      break;
    case 1:
      AnimateResizeFlip (x, y, w, h, fx, fy, fw, fh);
      break;
    case 2:
      AnimateResizeTurn (x, y, w, h, fx, fy, fw, fh);
      break;
    case 3:
      AnimateResizeZoom (x, y, w, h, fx, fy, fw, fh);
      break;
    default:
      AnimateResizeZoom3D (x, y, w, h, fx, fy, fw, fh);
      break;
    }
}

#if 0
/*
 * This makes an animation that looks like that light effect
 * when you turn off an old TV.
 * Used for window destruction
 */
void
AnimateClose (int x, int y, int w, int h)
{
  int i, step;

  if (h > 4)
    {
      step = h * 4 / Animate.iterations;
      if (step == 0)
	{
	  step = 2;
	}
      for (i = h; i >= 2; i -= step)
	{
	  XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, w, i);
	  XFlush (dpy);
	  sleep_a_little (ANIM_DELAY2 * 600);
	  XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, w, i);
	  y += step / 2;
	}
    }
  if (w < 2)
    return;
  step = w * 4 / Animate.iterations;
  if (step == 0)
    {
      step = 2;
    }
  for (i = w; i >= 0; i -= step)
    {
      XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, i, 2);
      XFlush (dpy);
      sleep_a_little (ANIM_DELAY2 * 1000);
      XDrawRectangle (dpy, Scr.Root, DrawGC, x, y, i, 2);
      x += step / 2;
    }
  sleep_a_little (100000);
  XFlush (dpy);
}
#endif

/* error_handler:
 * catch X errors, display the message and continue running.
 */
int
error_handler (Display * disp, XErrorEvent * event)
{
  fprintf (stderr, "%s: internal error, error code %d, request code %d, minor code %d.\n",
	 MyName, event->error_code, event->request_code, event->minor_code);
  return 0;
}

void
DeadPipe (int foo)
{
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	  "%s [--version|-v] [--help|-h]\n", MyName);
  exit (0);
}

int
main (int argc, char **argv)
{
  int i;
  char *global_config_file = NULL;
  unsigned long color;
  XGCValues gcv;

  /* Save our program name - for error messages */
  SetMyName (argv[0]);

  i = ProcessModuleArgs (argc, argv, &(global_config_file), NULL, NULL, usage);

  /* Dead pipe == AS died */
  signal (SIGPIPE, DeadPipe);
  signal (SIGQUIT, DeadPipe);
  signal (SIGSEGV, DeadPipe);
  signal (SIGTERM, DeadPipe);

  x_fd = ConnectX (&Scr, display_name, PropertyChangeMask);
  XSetErrorHandler (error_handler);

  /* connect to AfterStep */
  fd_width = GetFdWidth ();
  fd[0] = fd[1] = ConnectAfterStep (mask_reg);

  LoadConfig (global_config_file, "animate", ParseOptions);

  color = WhitePixel (dpy, screen);
  if (Animate.color)
    {
      XColor xcol;
      if (XParseColor (dpy, DefaultColormap (dpy, screen), Animate.color, &xcol))
	{
	  if (XAllocColor (dpy, DefaultColormap (dpy, screen), &xcol))
	    color = xcol.pixel;
	  else
	    color = 0;

	  if (color == 0)
	    {
	      fprintf (stderr, "\n%s: could not allocate color '%s' - using WHITE\n",
		       MyName, Animate.color);
	      color = WhitePixel (dpy, screen);
	    }

	}
      else
	{
	  fprintf (stderr, "%s: could not parse color '%s'\n",
		   MyName, Animate.color);
	}
    }

  gcv.line_width = Animate.width;
  gcv.function = GXxor;
  gcv.foreground = color;
  gcv.background = color;
  gcv.subwindow_mode = IncludeInferiors;
  DrawGC = XCreateGC (dpy, Scr.Root, GCFunction |
		      GCForeground |
		      GCLineWidth |
		      GCBackground |
		      GCSubwindowMode,
		      &gcv);
  SendInfo (fd, "Nop \"\"", 0);
  Loop ();
  return 0;
}

void
GetWindowSize (Window win, int *x, int *y, int *w, int *h)
{
  Window root_r;
  unsigned int bw_r, d_r;

  XGetGeometry (dpy, win, &root_r, x, y, (unsigned int *) w, (unsigned int *) h,
		&bw_r, &d_r);
}

/*
 * Wait for some event like iconify, deiconify and stuff.
 * 
 */
void
Loop ()
{
  int x0, y0, w0, h0, xf, yf, wf, hf;
  Window win;
  ASMessage *asmsg = NULL;

  while (1)
    {
      asmsg = CheckASMessage (fd[1], -1);
      if (asmsg)
	{
	  switch (asmsg->header[1])
	    {
	    case M_DEICONIFY:
	      win = (Window) * (asmsg->body + 1);
	      x0 = (int) *(asmsg->body + 3);
	      y0 = (int) *(asmsg->body + 4);

	      GetWindowSize (win, &xf, &yf, &wf, &hf);
	      Animate.resize (x0, y0, 64, 64, xf, yf, wf, hf);
	      DestroyASMessage (asmsg);
	      break;

	    case M_ICONIFY:
	      xf = (int) *(asmsg->body + 3);
	      yf = (int) *(asmsg->body + 4);
	      if (xf <= -10000)
		break;
	      wf = 64;
	      hf = 64;
	      DestroyASMessage (asmsg);

	      SendInfo (fd, "UNLOCK 1\n", 0);
	      asmsg = CheckASMessage (fd[1], -1);
	      if (!asmsg)
		break;
	      if (asmsg->header[1] != M_CONFIGURE_WINDOW)
		{
		  DestroyASMessage (asmsg);
		  break;
		}
	      x0 = (int) *(asmsg->body + 3);
	      y0 = (int) *(asmsg->body + 4);
	      w0 = (int) *(asmsg->body + 5);
	      h0 = (int) *(asmsg->body + 6);
	      Animate.resize (x0, y0, w0, h0, xf, yf, wf, hf);
	      DestroyASMessage (asmsg);
	      break;
	    default:
	      DestroyASMessage (asmsg);
	    }
	}
      SendInfo (fd, "UNLOCK 1\n", 0);
    }
}

/*****************************************************************************
 * 
 * This routine is responsible for reading and parsing the config file
 * Ripped from the Pager.
 *
 ****************************************************************************/
char *
strtrim (char *str, char **retbuf)
{
  char *beg, *end;

  beg = str;
  while (isspace (*beg))
    beg++;
  end = beg;
  while ((*end) && (!isspace (*end)) && (*end != '\n'))
    end++;
  strncpy (*retbuf, beg, end - beg);
  *(*retbuf + (end - beg)) = '\0';
  return (*retbuf);
}

void
ParseOptions (const char *filename)
{
  FILE *fd;
  char *line, *tline;
  char *parsebuf;
  unsigned int seed;
  long curtime;
  int len;

  if ((fd = fopen (filename, "r")) == NULL)
    {
      fprintf (stderr, "%s: can't open config file %s", MyName, filename);
      exit (1);
    }
  line = (char *) safemalloc (MAXLINELENGTH);
  parsebuf = (char *) safemalloc (MAXLINELENGTH);
  len = strlen (MyName);
  while (fgets (line, MAXLINELENGTH, fd) != NULL)
    {
      if ((*line == '*') && (!mystrncasecmp (line + 1, MyName, len)))
	{
	  tline = line + len + 1;
	  if (!mystrncasecmp (tline, "Color", 5))
	    {
	      strtrim (tline + 5, &parsebuf);
	      if (Animate.color)
		free (Animate.color);
	      Animate.color = strdup (parsebuf);
	    }
	  else if (!mystrncasecmp (tline, "Delay", 5))
	    {
	      Animate.delay = atoi (strtrim (tline + 5, &parsebuf));
	    }
	  else if (!mystrncasecmp (tline, "Width", 5))
	    {
	      Animate.width = atoi (strtrim (tline + 5, &parsebuf));
	    }
	  else if (!mystrncasecmp (tline, "Twist", 5))
	    {
	      Animate.twist = atof (strtrim (tline + 5, &parsebuf));
	    }
	  else if (!mystrncasecmp (tline, "Iterations", 10))
	    {
	      Animate.iterations = atoi (strtrim (tline + 10, &parsebuf));
	    }
	  else if (!mystrncasecmp (tline, "Resize", 6))
	    {
	      strtrim (tline + 6, &parsebuf);
	      if (!mystrncasecmp (parsebuf, "Twist", 5))
		Animate.resize = AnimateResizeTwist;
	      else if (!mystrncasecmp (parsebuf, "Flip", 4))
		Animate.resize = AnimateResizeFlip;
	      else if (!mystrncasecmp (parsebuf, "Turn", 4))
		Animate.resize = AnimateResizeTurn;
	      else if (!mystrncasecmp (parsebuf, "Zoom", 4))
		Animate.resize = AnimateResizeZoom;
	      else if (!mystrncasecmp (parsebuf, "Zoom3D", 6))
		Animate.resize = AnimateResizeZoom3D;
	      else if (!mystrncasecmp (parsebuf, "Random", 6))
		{
		  Animate.resize = AnimateResizeRandom;
		  curtime = time (NULL);
		  seed = (unsigned) curtime % INT_MAX;
		  srand (seed);
		}
	      else
		{
		  Animate.resize = AnimateResizeFlip;
		  fprintf (stderr, "%s: Bad Resize method '%s'\n", MyName, parsebuf);
		}
	    }
	}
    }

  free (parsebuf);
  free (line);
  fclose (fd);
}
